mlflow/pyfunc/__init__.py

```diff
@@ -656,15 +672,235 @@ def _warn_dependency_requirement_mismatches(model_path):
+
+    return _PYTHON_ENV_CACHE_DIR
+
+
+_PICKLE_PROTOCOL_FOR_RESTORE_PY_ENV = 3
+
+
+class PyFuncModelWithRestoredPythonEnv(PyFuncModel):
+
+    def __init__(
+            self,
+            model_meta: Model,
+            loader_module: str,
+            model_path: str,
+            py_version: str,
+            requirements_file_path: str,
+    ):
+        if not model_meta:
+            raise MlflowException("Model is missing metadata.")
+        self._model_meta = model_meta
+        self._loader_module = loader_module
+        self._model_path = model_path
+        self._py_version = py_version
+        self._requirements_file_path = requirements_file_path
+        self._subproc = None
+        self._virtual_env_python_bin = None
+        self._comm_sock = None
+
+    def _setup_restored_python_env_subproc(self):
+        env_cache_dir = _get_or_create_python_env_cache_dir()
+        pyenv_home_path = os.path.join(
+            env_cache_dir, 'pythons'
+        )
+        python_installation_path = os.path.join(
+            pyenv_home_path, '.pyenv', 'versions', self._py_version
+        )
+        if not os.path.exists(python_installation_path):
+            # install specified python version
+            sub_env = os.environ.copy()
+            sub_env['HOME'] = pyenv_home_path
+            subprocess.run(
+                [_PYENV_CMD, 'install', '-v', self._py_version],
+                check=True, stdout=sys.stdout, stderr=sys.stderr,
+                env=sub_env
+            )
+        else:
+            _logger.info(f'Found python installation cache at {python_installation_path}')
+        python_bin_path = os.path.join(
+            python_installation_path, 'bin/python'
+        )
+        req_file_content_sha = get_file_sha_hexdigest(self._requirements_file_path)
+        virtual_env_path = os.path.join(
+            env_cache_dir, 'venvs', req_file_content_sha
+        )
+        if not os.path.exists(virtual_env_path):
+            # install virtual env
+            subprocess.run(
+                [_VIRTUALENV_CMD, f'--python={python_bin_path}', virtual_env_path]
+            )
+            virtual_env_pip = os.path.join(
+                virtual_env_path, 'bin/pip'
+            )
+            subprocess.run(
+                [virtual_env_pip, 'install', '-r', self._requirements_file_path],
+                check=True, stdout=sys.stdout, stderr=sys.stderr
+            )
+            # For debug purpose, install the dev mlflow version in virtual env
+            subprocess.run(
+                [virtual_env_pip, 'install', '-e',
+                 os.path.dirname(os.path.dirname(mlflow.__file__))]
+            )
+        else:
+            _logger.info(f'Found python virtual env cache at {virtual_env_path}')
+
+        self._virtual_env_python_bin = os.path.join(
+            virtual_env_path, 'bin/python'
+        )
+        # setup tcp connection with subproc
+        listen_sock = socket.socket()
```

Comment:
The reason I use socket to transferring data between processes is:
* using stdin/stdout is risky, logger or some other routine might also write data to stdout.
* using pipe seems not possible because subprocess is not forked from parent process.

then socket become best choice. Also note spark python UDF implementation it also using socket to transfer data for sub-process executing python UDF routine.

Reply:
It's just a one-time hit to a relatively slow process since it's only going to transfer this environment a single time per execution session though, right? It seems like a smart design to me! :)